Forge Networking Remastered Unity/Assets/Bearded Man Studios Inc/Scripts/Networking/Forge/Networking/FacepunchP2PClient.cs

```diff
@@ -0,0 +1,453 @@
+﻿#if FACEPUNCH_STEAMWORKS
+using BeardedManStudios.Forge.Networking.Frame;
+using BeardedManStudios.Threading;
+using Steamworks;
+using System;
+using System.Net.Sockets;
+using System.Threading;
+
+namespace BeardedManStudios.Forge.Networking
+{
+	public class FacepunchP2PClient : BaseFacepunchP2P, IClient
+	{
+		/// <summary>
+		/// The max amount of tries that this client will attempt to connect to the server
+		/// where there is 3 seconds between each attempt
+		/// </summary>
+		public const int CONNECT_TRIES = 10;
+		public UDPPacketManager packetManager = new UDPPacketManager();
+		public event BaseNetworkEvent connectAttemptFailed;
+
+		/// <summary>
+		/// The identity of the server as a player
+		/// </summary>
+		private NetworkingPlayer server = null;
+		public NetworkingPlayer Server { get { return server; } }
+
+		/// <summary>
+		/// The hash that is / was validated by the server
+		/// </summary>
+		private string headerHash = string.Empty;
+
+		/// <summary>
+		/// Used to determine if the client has requested to be accepted by the server
+		/// </summary>
+		private bool headerExchanged = false;
+
+		/// <summary>
+		/// Callback for SteamNetworking.OnP2PConnectionFailed
+		/// </summary>
+		/// <param name="remoteSteamId">SteamId of the remote peer</param>
+		private void OnP2PConnectionFailed(SteamId remoteSteamId)
+		{
+			Logging.BMSLog.Log("OnP2PConnectionFailed called. Remote steamId: " + remoteSteamId.Value.ToString());
```

Comment:
Might be a good idea to introduce an event that people can subscribe to and be called from here?

Reply:
The user can do this from their own implementation, as the SteamNetworking class is static. I'll remove this debugging code.